src/main/java/com/nvidia/spark/rapids/jni/HyperLogLogPlusPlusHostUDF.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2024-2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.jni;
+
+import ai.rapids.cudf.ColumnVector;
+import ai.rapids.cudf.ColumnView;
+import ai.rapids.cudf.NativeDepsLoader;
+
+/**
+ * HyperLogLogPlusPlus(HLLPP) host UDF aggregation utils
+ */
+public class HyperLogLogPlusPlusHostUDF {
+  static {
+    NativeDepsLoader.loadNativeDeps();
+  }
+
+  /**
+   * HyperLogLogPlusPlus(HLLPP) aggregation types
+   */
+  public enum AggregationType {
+
+    /**
+     * Compute hash codes for the input, generate HyperLogLogPlusPlus(HLLPP)
+     * sketches from hash codes, and merge all the sketches into one sketch, output
+     * is a struct scalar with multiple long values.
+     */
+    Reduction(0),
+
+    /**
+     * Merge all HyperLogLogPlusPlus(HLLPP) sketches in the input column into one
+     * sketch. Input is a struct column with multiple long columns which is
+     * consistent with Spark. Output is a struct scalar with multiple long values.
+     */
+    ReductionMERGE(1),
+
+    /**
+     * Compute hash codes for the input, generate HyperLogLogPlusPlus(HLLPP)
+     * sketches from hash codes, and merge the sketches in the same group. Output is
+     * a struct column with multiple long columns which is consistent with Spark.
+     */
+    GroupBy(2),
+
+    /**
+     * Merge HyperLogLogPlusPlus(HLLPP) sketches in the same group.
+     * Input is a struct column with multiple long columns which is consistent with
+     * Spark.
+     */
+    GroupByMerge(3);
+
+    final int nativeId;
+
+    AggregationType(int nativeId) {
+      this.nativeId = nativeId;
+    }
+  }
+
+  /**
+   * Create a HyperLogLogPlusPlus(HLLPP) host UDF
+   */
+  public static long createHLLPPHostUDF(AggregationType type, int precision) {
```

Comment:
I think that this API leaks. We might want to think about how to redesign this and the java APIs to be more robust.

At a minimum we need an API that will let us free the HostUDF aggregation when we are done with it, but I really would prefer to have a class that acts as a wrapper around the long which is an AutoCloseable so we can use the withResource API in spark-rapids.

Reply:
Oh I forgot about closing it. Yes, we can auto-close the wrapper class https://github.com/rapidsai/cudf/blob/bbf4f7824c23c0c482f52bafdf1ece1213da2f65/java/src/main/java/ai/rapids/cudf/HostUDFWrapper.java#L28.
I'll add a fix for it shortly.

Reply:
Fix by https://github.com/rapidsai/cudf/pull/17727.

Reply:
Note that this class must extend `HostUPFWrapper` and override the required methods.

Reply:
done